zio-http/jvm/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala

```diff
@@ -106,38 +108,150 @@ private[netty] object NettyConnectionPool {
+
+    // Alternate between families
+    var useIpv6 = true
+    while (ipv6Iter.hasNext || ipv4Iter.hasNext) {
+
+      if (useIpv6 && ipv6Iter.hasNext) {
+        builder += ipv6Iter.next()
+      } else if (ipv4Iter.hasNext) {
+        builder += ipv4Iter.next()
+      } else if (ipv6Iter.hasNext) {
+        builder += ipv6Iter.next()
+      }
+
+      useIpv6 = !useIpv6
+    }
+
+    builder.result()
+  }
+
+  /**
+   * Implements Happy Eyeballs (RFC 8305) connection algorithm. Races connection
+   * attempts to IPv6 and IPv4 addresses with staggered delays.
+   */
+  private def happyEyeballsConnect(
+    resolvedHosts: Chunk[InetAddress],
+    channelFactory: JChannelFactory[JChannel],
+    eventLoopGroup: JEventLoopGroup,
+    location: URL.Location.Absolute,
+    initializer: ChannelInitializer[JChannel],
+    connectionTimeout: Option[Duration],
+    localAddress: Option[InetSocketAddress],
+  )(implicit trace: Trace): ZIO[Scope, Throwable, JChannel] = {
+
+    if (resolvedHosts.isEmpty) {
+      ZIO.fail(new RuntimeException("No addresses to connect to"))
+    } else if (resolvedHosts.size == 1) {
+      connectToAddress(
+        resolvedHosts.head,
+        channelFactory,
+        eventLoopGroup,
+        location,
+        initializer,
+        connectionTimeout,
+        localAddress,
+      )
+    } else {
+      val addresses = sortAddresses(resolvedHosts)
+      for {
+        lastFailed <- Queue.bounded[Unit](requestedCapacity = 1)
+        successful <- Ref.make(List.empty[JChannel])
+        _          <- ZIO.raceAll(
+          connectToAddress(
+            addresses.head,
+            channelFactory,
+            eventLoopGroup,
+            location,
+            initializer,
+            connectionTimeout,
+            localAddress,
+          ).onExit {
+            case e: Exit.Success[JChannel] => successful.update(channels => channels :+ e.value)
+            case _: Exit.Failure[_]        => lastFailed.offer(())
+          },
+          addresses.tail.zipWithIndex.map { case (address, index) =>
+            ZIO.sleep(HappyEyeballsDelay * index.toDouble).raceFirst(lastFailed.take).ignore *>
+              connectToAddress(
+                address,
+                channelFactory,
+                eventLoopGroup,
+                location,
+                initializer,
+                connectionTimeout,
+                localAddress,
+              ).onExit {
+                case e: Exit.Success[JChannel] => successful.update(channels => channels :+ e.value)
+                case _: Exit.Failure[_]        => lastFailed.offer(())
+              }
+          },
+        )
+        channels   <- successful.get
+        channel    <- channels.headOption match {
+          case ch: Some[JChannel] => ZIO.succeed(ch.value)
+          case None               => ZIO.fail(new RuntimeException("All connection attempts failed"))
+        }
+        _          <- ZIO.foreachDiscard(channels.tail)(ch => ZIO.attempt(ch.close()))
```

Comment:
There's a potential leak here again in case that closing one of them fails. I think this is better:

```suggestion
        _          <- ZIO.foreachDiscard(channels.tail)(ch => ZIO.ignore(ch.close()))
```